Run clang as a standalone binary during runtime compilation

ISauve opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Replaces runtime compilation via direct interface with the clang API with runtime compilation via invocation of the clang & llc binaries.

### Motivation

During runtime compilation, clang makes a large number of allocations using an LLVM allocator which does not release its memory upon destruction, only upon process exit. This was causing a large increase in system-probe memory usage when runtime compilation was enabled. By invoking the clang binary to perform runtime compilation, we ensure that the memory it allocates is released upon compilation completion.

### Additional Notes

**System-probe memory usage with runtime compilation enabled (no kernel header downloading)**

Before: 
![RC_A](https://user-images.githubusercontent.com/11351618/152474291-90715a05-9085-486e-ae06-4fb16ac66e81.png)

After:
![RC_B](https://user-images.githubusercontent.com/11351618/152474300-24197c55-3121-44f1-b930-68a6c35bae64.png)

**System-probe memory usage with runtime compilation enabled (with kernel header downloading)**

Before: 
![RC_C](https://user-images.githubusercontent.com/11351618/152474306-d470ec0a-d9f1-4c2b-8a6d-9fb3376822f4.png)

After:
![RC_D](https://user-images.githubusercontent.com/11351618/152474316-bbaf191b-1310-404d-96d7-16c17df873fa.png)

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Enable all runtime compiled objects:
```
system_probe_config:
  log_level: debug
  enable_runtime_compiler: true
  enable_tcp_queue_length: true
  enable_oom_kill: true
runtime_security_config:
  enabled: true
  enable_runtime_compiled_constants: true
network_config:
  enabled: true
  enable_http_monitoring: true
```
Run the system-probe and search the logs for `successfully compiled runtime version of`. You should see 7 instances:
```
successfully compiled runtime version of tracer.c
successfully compiled runtime version of conntrack.c
successfully compiled runtime version of http.c
successfully compiled runtime version of tcp-queue-length.c
successfully compiled runtime version of oom-kill.c
successfully compiled runtime version of constant_fetcher.c
successfully compiled runtime version of runtime-security.c
```

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ISauve committed:
Include clang & llc binaries & stdarg.h in agent package

ISauve committed:
Add more logging + capture http compilation telemetry

ISauve committed:
Compile using clang & llc binaries instead of clang API

ISauve committed:
Update build_dev_image command to package clang, llc & stdarg.h

ISauve committed:
Ignore int-conversion warnings when compiling oomkill & tcp queue length checks

ISauve committed:
Set target when compiling http.c

ISauve committed:
Merge branch 'main' into isabelle.sauve/RC-memory-fix

ISauve committed:
Fix gofmt

ISauve committed:
Set clang & llc binaries to 755 permission

ISauve committed:
Fix http compilation on arm

nplanel commented on .gitlab/source_test/ebpf.yml:
```diff
@@ -1,8 +1,12 @@
+  - mkdir -p $DATADOG_AGENT_EMBEDDED_PATH/bin
+  - mkdir -p $DATADOG_AGENT_EMBEDDED_PATH/include
+  - $S3_CP_CMD $S3_PERMANENT_ARTIFACTS_URI/clang.$ARCH $DATADOG_AGENT_EMBEDDED_PATH/bin/clang
+  - $S3_CP_CMD $S3_PERMANENT_ARTIFACTS_URI/clang-11.$ARCH $DATADOG_AGENT_EMBEDDED_PATH/bin/clang-11
```
do we really need to have a versioned version ? clang-<major> here ?
or clang call itself clang-<major> ?

brycekahle replied:
I agree that we only need a single `clang` file, so renaming the binary to `clang` when copying from S3 during the package builds makes the most sense.

nplanel commented on pkg/network/http/compile.go:
```diff
@@ -22,6 +24,13 @@ func getRuntimeCompiledHTTP(config *config.Config) (runtime.CompiledOutput, erro
 func getCFlags(config *config.Config) []string {
 	var cflags []string
+
+	if strings.ToLower(rt.GOARCH) == "arm64" {
```
may just fixup only arch that need to
```
arch := strings.ToLower(rt.GOARCH)
if arch ==  "arm64" {
  arch = "aarch64"
}
cflags = append(cflags, fmt.Sprintf("--target=%s-unknown-linux-gnu", arch))
```

brycekahle replied:
I think the `strings.ToLower` is unnecessary. Here are all possible values: https://github.com/golang/go/blob/master/src/go/build/syslist.go#L11

hmahmood replied:
We should check for `arm64` and `amd64` here if we want to be explicit and error out in out in other cases. Bit overkill, but I think more future proof.

brycekahle commented on .gitlab/deps_build.yml:
```diff
@@ -31,16 +31,12 @@
   artifacts:
     expire_in: 2 weeks
-    paths:
-      - $CI_PROJECT_DIR/.tmp/clang-$ARCH-11.0.1.tar.xz
```
It would be nice to include the two binaries here, since it makes it easy to download them for use in our agent VM.

brycekahle commented on pkg/collector/corechecks/ebpf/probe/oom_kill.go:
```diff
@@ -40,7 +40,7 @@ type OOMKillProbe struct {
 
 func NewOOMKillProbe(cfg *ebpf.Config) (*OOMKillProbe, error) {
-	compiledOutput, err := runtime.OomKill.Compile(cfg, nil)
+	compiledOutput, err := runtime.OomKill.Compile(cfg, []string{"-Wno-error=int-conversion"})
```
Just curious, do you know what lines are causing these errors?

hmahmood replied:
These really should not be ignored if at all possible.

ISauve replied:
For the oom kill check:
https://github.com/DataDog/datadog-agent/blob/729162b115e91331040de46450626cdd28ef1ece/pkg/collector/corechecks/ebpf/c/runtime/oom-kill-kern.c#L34

For the tcp queue length check:
https://github.com/DataDog/datadog-agent/blob/729162b115e91331040de46450626cdd28ef1ece/pkg/collector/corechecks/ebpf/c/runtime/tcp-queue-length-kern.c#L106 and https://github.com/DataDog/datadog-agent/blob/729162b115e91331040de46450626cdd28ef1ece/pkg/collector/corechecks/ebpf/c/runtime/tcp-queue-length-kern.c#L125

If it's an issue, I can remove this flag & just cast the return values of `PT_REGS_PARM1` to solve this error, that's fine too

brycekahle commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -8,134 +8,59 @@
 	arch := kernel.Arch()
 	if arch == "" {
-		return nil, fmt.Errorf("unable to get kernel arch for %s", runtime.GOARCH)
+		fmt.Errorf("unable to get kernel arch for %s", runtime.GOARCH)
```
```suggestion
		return fmt.Errorf("unable to get kernel arch for %s", runtime.GOARCH)
```

brycekahle commented on pkg/ebpf/compiler/compiler_test.go:
```diff
@@ -31,26 +30,22 @@ func TestCompilerMatch(t *testing.T) {
 		}
 		return
 	}
+	input, err := os.Open(cPath)
+	require.NoError(t, err)
```
```suggestion
	input, err := os.Open(cPath)
	require.NoError(t, err)
	defer input.Close()
```

brycekahle commented on .gitlab/deps_build.yml:
```diff
@@ -31,16 +31,12 @@
-    - mkdir -p $CI_PROJECT_DIR/.tmp
-    - tar cvaf $CI_PROJECT_DIR/.tmp/clang-$ARCH-11.0.1.tar.xz .
-    - $S3_CP_CMD $CI_PROJECT_DIR/.tmp/clang-$ARCH-11.0.1.tar.xz $S3_PERMANENT_ARTIFACTS_URI/clang-$ARCH-11.0.1.tar.xz
+    - $S3_CP_CMD $DATADOG_AGENT_EMBEDDED_PATH/bin/clang $S3_PERMANENT_ARTIFACTS_URI/clang.${ARCH} 
```
We should include the full version in the S3 path, so that when we upgrade versions we don't clobber old artifacts that could still be in use.

brycekahle commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -147,7 +72,44 @@ func NewEBPFCompiler(headerDirs []string, verbose bool) (*EBPFCompiler, error) {
+
+	bcToObj := exec.CommandContext(llcCtx, llcBinPath, "-march=bpf", "-filetype=obj", "-o", outputFile, "-")
+	bcToObj.Stdin = &clangOut
+	bcToObj.Stdout = &llcOut
```
If we don't actually care about `stdout` for error reporting, then this can be `nil` instead.

brycekahle reviewed (changes requested):
Commented on several things, but reinstating the flags hash and full versioned S3 artifacts are necessary.

ISauve committed:
Rename clang-11 to clang & add version to s3 upload path

ISauve committed:
Add back flag hash to output file path

ISauve committed:
Add missing returns to CompileToObjectFile

ISauve committed:
Close input during compiler_test.go

ISauve committed:
Wrap glibc symbols to older version when building clang

ISauve committed:
Improve compilation error messages

ISauve committed:
Remove Wno-error=int-conversion flag from oomkill/tcpqlength

ISauve committed:
Small fixes

ISauve committed:
Fix imports

brycekahle commented on pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go:
```diff
@@ -12,35 +12,13 @@ import (
 
 	"github.com/DataDog/datadog-agent/pkg/ebpf"
 	"github.com/DataDog/datadog-agent/pkg/ebpf/compiler"
+	"github.com/DataDog/datadog-agent/pkg/security/log"
```
I don't think this is the correct `log` import. It should be `"github.com/DataDog/datadog-agent/pkg/util/log"`

ISauve replied:
Oops, good catch!

brycekahle commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -8,134 +8,59 @@
 	}
 
-	var cflags []string
+	cflags = append(cflags, defaultFlags...)
```
Moving the default flags here means they won't be included in the flags hash. If we decide to change the default flags, but the content of a `.c` file doesn't get updated, then it wouldn't be recompiled. Can we move the defaults back to their previous location?

brycekahle replied:
I also don't think this is preferred from a design standpoint. We should have the flexibility to provide exactly what flags will be used. There may be cases where the defaults are not preferable, and we should be able to override that.

ISauve committed:
Include default compilation flags in hash

ISauve commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -110,7 +110,7 @@ func CompileToObjectFile(in io.Reader, outputFile string, cflags []string, heade
 	if err != nil {
 		var errMsg string
-		if clangCtx.Err() == context.DeadlineExceeded {
+		if llcCtx.Err() == context.DeadlineExceeded {
```
Yes, that was the intent.

ISauve committed:
Change clang & system-probe permissions to 0744

ISauve committed:
Verify that clang/llc only target bpf

ISauve committed:
Simplify diff command for checking clang/llc targets

ISauve committed:
Fix cloning issue for arm builder

ISauve committed:
Move patch before cmake setup command

ISauve committed:
Merge branch 'main' into isabelle.sauve/RC-memory-fix

ISauve committed:
Fix kitchen tests

ISauve committed:
Update tests

ISauve committed:
Fix passing of flags to compilation function

ISauve committed:
Merge branch 'main' into isabelle.sauve/RC-memory-fix

brycekahle commented on pkg/network/http/compile.go:
```diff
@@ -22,6 +23,13 @@ func getRuntimeCompiledHTTP(config *config.Config) (runtime.CompiledOutput, erro
 func getCFlags(config *config.Config) []string {
 	var cflags []string
+
+	if rt.GOARCH == "arm64" {
```
In the current state of `main`, I don't think this is necessary since we reverted the PR that needed it.

ISauve replied:
Fair enough, I'll remove this for now & we can add it back along with the relevant change in the future

brycekahle commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -147,7 +51,52 @@ func NewEBPFCompiler(headerDirs []string, verbose bool) (*EBPFCompiler, error) {
+	compileToBC := exec.CommandContext(clangCtx, clangBinPath, cflags...)
+	compileToBC.Stdin = in
+	compileToBC.Stdout = &clangOut
+	compileToBC.Stderr = &clangErr
```
Could we log the `stderr` from `clang` and `llc` as a debug log message? I think that could be helpful debugging issues.

ISauve replied:
If `stderr` is not nil, we return it from this function, so I don't think logging it would add anything here

brycekahle replied:
We only do that if `err != nil`. I'm looking for a debug log when compilation completes successfully.

ISauve replied:
Ah, I see. We can definitely do that.

brycekahle commented on pkg/ebpf/c/stdarg.h:
```diff
@@ -1,4 +1,3 @@
-R"***(
 /*===---- stdarg.h - Variable argument handling ----------------------------===
```
I don't think we should leave this file in the `pkg/ebpf/c` since those files are meant to be used directly by ebpf programs. Being underneath `pkg/ebpf/compiler` makes more sense to me.

We also have this file in-tree, as well as a version coming from S3. Since this file is Apache2 licensed, we should be able to embed it. Could we try using `go:embed` and then writing it to a temp file during compilation? That would be one less file we have to worry about in our packaging.

ISauve replied:
that's a great idea!

brycekahle commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -147,7 +51,52 @@ func NewEBPFCompiler(headerDirs []string, verbose bool) (*EBPFCompiler, error) {
+	defer clangCancel()
 
-	return ebpfCompiler, nil
+	compileToBC := exec.CommandContext(clangCtx, clangBinPath, cflags...)
```
Can we add a debug log of the command being run for `clang` and `llc`?

ISauve committed:
Log the clang/llc commands at debug level

ISauve committed:
Move stdarg.h to pkg/ebpf/compiler & embed it

ISauve committed:
Remove unnecessary flags from http compilation

ISauve committed:
Update runtime compiled asset hashes

ISauve committed:
Remove extra 'end' statement from kitchen test file

brycekahle commented on .gitlab/deps_build.yml:
```diff
@@ -31,16 +44,26 @@
+    # fail if clang or llc targets anything other than bpf
+    - $DATADOG_AGENT_EMBEDDED_PATH/bin/clang-11 -print-targets > clang_targets.txt
+    - diff $SRC_PATH/pkg/ebpf/c/expected_targets.txt clang_targets.txt
+    - $DATADOG_AGENT_EMBEDDED_PATH/bin/llc --version > llc_targets.txt
+    - sed -i -e 1,6d llc_targets.txt # remove first 6 lines of llc version output; we only care about the registered targets
```
```suggestion
    - $DATADOG_AGENT_EMBEDDED_PATH/bin/llc --version | tail +7 > llc_targets.txt # remove first 6 lines of llc version output; we only care about the registered targets
```

ISauve committed:
Import embed package with blank identifier

ISauve committed:
Decrease clang build size

ISauve committed:
Log clang/llc stderr as debug log messages

ISauve committed:
Combine deps_build commands

ISauve committed:
Merge branch 'main' into isabelle.sauve/RC-memory-fix

brycekahle commented on .gitlab/deps_build.yml:
```diff
@@ -29,18 +42,28 @@
+    - diff $SRC_PATH/pkg/ebpf/c/expected_targets.txt clang_targets.txt
+    - $DATADOG_AGENT_EMBEDDED_PATH/bin/llc --version | tail -n +7 > llc_targets.txt # remove first 6 lines of llc version output; we only care about the registered targets
+    - diff $SRC_PATH/pkg/ebpf/c/expected_targets.txt llc_targets.txt
+    - $S3_CP_CMD $DATADOG_AGENT_EMBEDDED_PATH/bin/clang-11 $S3_PERMANENT_ARTIFACTS_URI/clang-11.0.1.${ARCH}
```
It has been requested that we name these binaries `clang-bpf` and `llc-bpf`, so the intent is a little clearer to our customers. Can you make that change?

ISauve committed:
Merge branch 'main' into isabelle.sauve/RC-memory-fix

ISauve committed:
Append -bpf to clang and llc binary names

ISauve committed:
Fix Wunused-function error when compiling on older kernels

ISauve commented on pkg/network/ebpf/c/runtime/tracer.c:
```diff
@@ -31,23 +31,6 @@
 # error "kernel version not included?"
 #endif
 
-static __always_inline __be32 rt_nexthop_bpf(struct rtable *rt) {
-    if (!rt) {
-        return 0;
-    }
-    __be32 hop = 0;
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 2, 0)
-    bpf_probe_read(&hop, sizeof(hop), &rt->rt_gateway);
-#else
-    u8 family;
-    bpf_probe_read(&family, sizeof(family), &rt->rt_gw_family);
-    if (family == AF_INET) {
-        bpf_probe_read(&hop, sizeof(hop), &rt->rt_gw4);
-    }
-#endif
-    return hop;
-}
-
```
Note: I removed this because it's currently unused

ISauve committed:
Enforce that we can only load root-writeable object files

ISauve committed:
Ensure we only load root-writeable ebpf programs

ISauve committed:
Merge branch 'main' into isabelle.sauve/RC-memory-fix

ISauve committed:
Copy bpf files out of mounted directory

ISauve committed:
Skip permissions check for embedded files

brycekahle commented on pkg/network/ebpf/c/sockfd.h:
```diff
@@ -35,7 +35,7 @@ struct bpf_map_def SEC("maps/pid_fd_by_sock") pid_fd_by_sock = {
 };
 
-static __always_inline void clear_sockfd_maps(struct sock* sock) {
+__attribute__((unused)) static __always_inline void clear_sockfd_maps(struct sock* sock) {
```
❓ I see this used in `kprobe/tcp_close`

ISauve replied:
This mitigates an issue that can occur on older kernels where `clang` can generate `Wunused-function` warnings for static inline functions defined in header files (even if they end up being used in a source file later on) -  see [this failed job](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/125110047#L738) as an example.

I'll change the attribute to `maybe_unused` & leave some comments where this is done to make it more clear

brycekahle replied:
I think those functions are actually unused in that particular `.c` file. For example: `flip_tuple` is only used in `http.c`.

We can use `maybe_unused` for now. We should have a discussion about if we want to restructure the header files to avoid this.

ISauve commented:
> Does this work for embedded files?

Good catch - I tried this out with the `bundle_ebpf` build tag & (unsurprisingly) it doesn't work for embedded files. I've removed the check on that code path

brycekahle commented on tasks/system_probe.py:
```diff
@@ -680,6 +675,14 @@ def build_object_files(ctx, parallel_build):
     generate_runtime_files(ctx)
 
+    # We need to copy the bpf files out of the mounted build directory in order to be able to
+    # change their ownership to root
```
Why can't we change them in-tree?

brycekahle replied:
Got it. Yay vagrant.

brycekahle replied:
Let's make sure we update the agent dev VM to not symlink these directories.

ISauve committed:
Add comments & swap __attribute__((unused)) with __maybe_unused

ISauve committed:
Use VerifyAssetIsRootWriteable helper function

ISauve committed:
Fix formatting

ISauve committed:
Fix system_probe.build task

ISauve committed:
Add VerifyAssetPermissions function for non-bpf builds

ISauve committed:
Set correct bpf file permissions

ISauve committed:
Fix preprocess test

derekwbrown committed:
Fixes unit tests for branch (#11901)

* Try changing default mode in kitchen test to 644
Temporarily add asset path to failure

fix glob

Fix missing tag on for windows
Try fixing up permissions in rspec rather than recipe

modify test to pass

skip http section on non-supported kernels

Try changing default mode in kitchen test to 644
Temporarily add asset path to failure

fix glob

Fix missing tag on for windows
Try fixing up permissions in rspec rather than recipe

modify test to pass

set log message back to previous state

* cherry-pick Fix TestGetStats  (#11732) to make
tests sane

* try fix line endings

* change tabs back to spaces

Co-authored-by: Hasan Mahmood <hmahmood@users.noreply.github.com>

derekwbrown committed:
Merge branch 'main' into isabelle.sauve/RC-memory-fix

derekwbrown committed:
fix multiple declaration

derekwbrown committed:
fix another merge mistake; no longer have dnsSupported() it tracer tests

derekwbrown committed:
fix additional test failure

brycekahle commented on .gitlab/package_build/deb.yml:
```diff
@@ -34,7 +34,10 @@
-    - chmod 755 /tmp/system-probe/system-probe
+    - $S3_CP_CMD $S3_PERMANENT_ARTIFACTS_URI/clang-11.0.1.${PACKAGE_ARCH} /tmp/system-probe/clang-bpf
+    - $S3_CP_CMD $S3_PERMANENT_ARTIFACTS_URI/llc-11.0.1.${PACKAGE_ARCH} /tmp/system-probe/llc-bpf
+    - chmod 0644 /tmp/system-probe/*.o
```
I see `deb` has this, but the other package builds don't? Should they be in sync?

ISauve committed:
Set correct asset permissions in all package builds

brycekahle commented on pkg/ebpf/bytecode/asset_windows.go:
```diff
@@ -0,0 +1,26 @@
+
+// VerifyAssetPermissions is for verifying the permissions of bpf programs
+func VerifyAssetPermissions(assetPath string) error {
+	return fmt.Errorf("verification of bpf assets is not supported on windows")
```
Is this necessary? There are no ebpf assets on windows

ISauve replied:
It needs to be defined on windows because it's used here: https://github.com/DataDog/datadog-agent/blob/18184094e46350b5b0e5ee8bcd0d4a3a67611890/pkg/ebpf/bytecode/asset_nobindata.go#L21

and there isn't a non-windows build tag protecting that file. Since it's all ebpf-related code, we could add the `!windows` build tag there & wherever that function is used if we want

brycekahle replied:
We weren't ever calling `GetReader` from windows, so it didn't matter (and this file didn't exist before and everything was OK). Has that changed?

ISauve replied:
What's changed is that `VerifyAssetPermissions` contains code that doesn't compile on windows: https://github.com/DataDog/datadog-agent/blob/18184094e46350b5b0e5ee8bcd0d4a3a67611890/pkg/ebpf/bytecode/asset.go#L33

So we need a windows version & a non-windows version, even if `GetReader` is never called from windows

brycekahle replied:
Can you create an `asset_linux.go` and move the `VerifyAssetPermissions` there? That way you don't have to redefine `AssetReader` here, and can remove the build tags from `asset.go`

ISauve replied:
Sure!

ISauve committed:
Move VerifyAssetPermissions to its own file

ISauve committed:
Remove io package
